sycl/doc/extensions/experimental/SYCL_EXT_ONEAPI_GRAPH.asciidoc

```diff
@@ -0,0 +1,290 @@
+
+== Introduction
+
+This extension introduces an interface that enables a lazy execution and easy replay of a kernel graph by separating
+Its definition and execution. 
+
+== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros".  Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_GRAPH` to one of the values defined in the table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's APIs the implementation supports.
+
+Table 1. Values of the `SYCL_EXT_ONEAPI_GRAPH` macro.
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1     |Initial extension version. Base features are supported.
+|===
+
+== SYCL Graph Terminology
+
+Table 2. Terminology.
+|===
+|Concept|Description
+|graph| Class that stores structured work units and their dependencies
+|node| The unit of work. Can have different attributes.
+|edge| Dependency between work units. Happens before relation.
+|===
+
+== Node
+
+Node is a class that can encapsulate SYCL kernel functions or host tasks for deferred execution.
+A graph has to be created first, the structure of a graph is defined second by adding nodes and edges.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+	class node{
+	};
+}
+----
+
+NOTE:
+
+== Edge
+
+A dependency between two nodes representing a happens before relationship.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+	// Adding dependency between two nodes.
+    	void make_edge(node sender, node receiver);
+}
+----
+
+== Graph
+
+Graph is a class that represents a directed acyclic graph of nodes. 
+A graph can be nested, can have multiple root nodes that are scheduled for execution first and multiple leaf nodes that are scheduled for execution last.
+Member functions as listed in Table 2 and 3 can be used to add nodes to a graph.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  class graph {
+  };
+
+}
+----
+
+=== Executable Graph
+
+`executable_graph` represents a user generated device and context specific execution object that can be submitted to a queue for execution. 
+The structure of an `executable_graph` object, such as adding nodes or edges, can not be changed.
+Each `executable_graph` object can only be executed once at the same time on its assigned queue.
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+  class executable_graph {
+  };
+
+}
+----
+
+
+Table 3. Constructors of the `graph` class.
+|===
+|Constructor|Description
+
+|`graph()`
+|Creates a `graph` object
+
+|`graph(graph& parent)`
+|Creates a nested `graph` object
+
+|===
+
+Table 4. Member functions of the `graph` class.
+|===
+|Member function|Description
+
+|`node add_empty_node(const std::vector<node>& dep = {});`
```

Comment:
I'm not a big fan of arguments like this, because the meaning of something like `add_empty_node({A, B})` in code is not obvious without referring to the documentation.

Adding a new type like we did for `reduction` might be nicer, here, something like:
```c++
graph.add_empty_node(sycl::depends_on(A, B));
```
...extending to other functions too:
```c++
auto node_c = g.add_device_node(
    [&](sycl::handler &h) {
      h.parallel_for(sycl::range<1>{n},
                     sycl::reduction(dotp, 0.0f, std::plus()),
                     [=](sycl::id<1> it, auto &sum) {
                       const size_t i = it[0];
                       sum += x[i] * z[i];
                     });
    },
    sycl::depends_on(node_a, node_b));
```

Reply:
Thanks for that suggestion. Looks like a good alternative solution for passing dependencies. My main concerns are increasing verbosity and that it is not in-line with how dependencies on `sycl::event` are passed with the default eager API. Another option then passing a `list` that is empty by default like in the current proposal, would be providing overloads for empty, single and multiple dependencies.
I would consider the `make_edge` function an alternative way of expressing dependencies where the meaning is kind of obvious.

Reply:
done. API has been slightly modified such that dependencies are passed via property list as a property. Thanks @Pennycook for the suggestion.